apps/app_rpt.c

```diff
@@ -1335,101 +1336,112 @@ void *rpt_call(void *this)
+		if (!channel_ref && (myrpt->callmode != CALLMODE_FAILED)) {
 			/* If patch is setup for far end disconnect */
 			if (myrpt->patchfarenddisconnect || (myrpt->p.duplex < 2)) {
-				ast_debug(1, "callmode=%i, patchfarenddisconnect=%i, duplex=%i\n",
-					myrpt->callmode, myrpt->patchfarenddisconnect, myrpt->p.duplex);
+				ast_debug(1,
+					"callmode=%i, patchfarenddisconnect=%i, duplex=%i\n",
+					myrpt->callmode,
+					myrpt->patchfarenddisconnect,
+					myrpt->p.duplex);
 				myrpt->callmode = CALLMODE_DOWN;
 				myrpt->macropatch = 0;
 				if (!myrpt->patchquiet) {
 					rpt_mutex_unlock(&myrpt->lock);
 					rpt_telemetry(myrpt, TERM, NULL);
 					rpt_mutex_lock(&myrpt->lock);
 				}
-			} else {			/* Send congestion until patch is downed by command */
+				continue;
+			} else { /* Send congestion until patch is downed by command */
 				myrpt->callmode = CALLMODE_FAILED;
 				rpt_mutex_unlock(&myrpt->lock);
 				/* start congestion tone */
 				rpt_play_congestion(genchannel);
 				rpt_mutex_lock(&myrpt->lock);
 			}
 		}
-		if (ast_channel_is_bridged(mychannel) && ast_channel_state(mychannel) == AST_STATE_UP)
-			if ((!sentpatchconnect) && myrpt->p.patchconnect && ast_channel_is_bridged(mychannel)
-				&& (ast_channel_state(mychannel) == AST_STATE_UP)) {
-				sentpatchconnect = 1;
-				rpt_mutex_unlock(&myrpt->lock);
-				rpt_telemetry(myrpt, PLAYBACK, (char*) myrpt->p.patchconnect);
-				rpt_mutex_lock(&myrpt->lock);
-			}
+		ast_channel_lock(channel_ref);
+		if ((!sentpatchconnect) && myrpt->p.patchconnect && ast_channel_is_bridged(channel_ref) &&
+			(ast_channel_state(channel_ref) == AST_STATE_UP)) {
+			ast_channel_unlock(channel_ref);
+			sentpatchconnect = 1;
+			rpt_mutex_unlock(&myrpt->lock);
+			rpt_telemetry(myrpt, PLAYBACK, (char *) myrpt->p.patchconnect);
+			rpt_mutex_lock(&myrpt->lock);
+		} else {
+			ast_channel_unlock(channel_ref);
+		}
 		if (myrpt->mydtmf) {
 			struct ast_frame wf = { AST_FRAME_DTMF, };
 
 			wf.subclass.integer = myrpt->mydtmf;
-			if (ast_channel_is_bridged(mychannel) && ast_channel_state(mychannel) == AST_STATE_UP) {
+			ast_channel_lock(channel_ref);
+			if (ast_channel_is_bridged(channel_ref) && ast_channel_state(channel_ref) == AST_STATE_UP) {
+				ast_channel_unlock(channel_ref);
 				rpt_mutex_unlock(&myrpt->lock);
-				ast_queue_frame(mychannel, &wf);
+				ast_queue_frame(channel_ref, &wf);
 				ast_senddigit(genchannel, myrpt->mydtmf, 0);
 				rpt_mutex_lock(&myrpt->lock);
+			} else {
+				ast_channel_unlock(channel_ref);
 			}
 			myrpt->mydtmf = 0;
 		}
+		ast_channel_unref(channel_ref);
 		rpt_mutex_unlock(&myrpt->lock);
 		usleep(MSWAIT * 1000);
 		rpt_mutex_lock(&myrpt->lock);
 	}
-	ast_debug(1, "exit channel loop\n");
+	ast_debug(1, "exit channel loop mode %d\n", myrpt->callmode);
 	rpt_mutex_unlock(&myrpt->lock);
 	rpt_stop_tone(genchannel);
-	if (ast_channel_pbx(mychannel))
-		ast_softhangup(mychannel, AST_SOFTHANGUP_DEV);
+	channel_ref = ast_channel_get_by_name(mychannel_name);
```

Comment:
I think this is a "best" practice for deciding to `softhangup()` or not.
Move forward with this part?